gapis/gfxapi/vulkan/dependency_graph.go

```diff
@@ -0,0 +1,1733 @@
+			for j := uint64(0); j < stageCount; j++ {
+				shaderStage := shaderStages.Index(j, s).Read(ctx, a, s, nil)
+				module := shaderStage.Module
+				addRead(&b, g, vulkanStateKey(module))
+			}
+			// read renderpass
+			renderPass := createInfos.Index(i, s).Read(ctx, a, s, nil).RenderPass
+			addRead(&b, g, vulkanStateKey(renderPass))
+			// Create pipeline
+			pipeline := pipelines.Index(i, s).Read(ctx, a, s, nil)
+			addWrite(&b, g, vulkanStateKey(pipeline))
+		}
+
+	case *RecreateGraphicsPipeline:
+		createInfo := typedAtom.PCreateInfo.Read(ctx, a, s, nil)
+		stageCount := uint64(createInfo.StageCount)
+		shaderStages := createInfo.PStages.Slice(0, stageCount, s)
+		for i := uint64(0); i < stageCount; i++ {
+			shaderStage := shaderStages.Index(i, s).Read(ctx, a, s, nil)
+			addRead(&b, g, vulkanStateKey(shaderStage.Module))
+		}
+		addRead(&b, g, vulkanStateKey(createInfo.RenderPass))
+		addWrite(&b, g, vulkanStateKey(typedAtom.PPipeline.Read(ctx, a, s, nil)))
+
+	case *VkCreateComputePipelines:
+		pipelineCount := uint64(typedAtom.CreateInfoCount)
+		createInfos := typedAtom.PCreateInfos.Slice(0, pipelineCount, s)
+		pipelines := typedAtom.PPipelines.Slice(0, pipelineCount, s)
+		for i := uint64(0); i < pipelineCount; i++ {
+			// read shader
+			shaderStage := createInfos.Index(i, s).Read(ctx, a, s, nil).Stage
+			module := shaderStage.Module
+			addRead(&b, g, vulkanStateKey(module))
+			// Create pipeline
+			pipeline := pipelines.Index(i, s).Read(ctx, a, s, nil)
+			addWrite(&b, g, vulkanStateKey(pipeline))
+		}
+
+	case *RecreateComputePipeline:
+		createInfo := typedAtom.PCreateInfo.Read(ctx, a, s, nil)
+		module := createInfo.Stage.Module
+		addRead(&b, g, vulkanStateKey(module))
+		addWrite(&b, g, vulkanStateKey(typedAtom.PPipeline.Read(ctx, a, s, nil)))
+
+	case *VkCreateShaderModule:
+		addWrite(&b, g, vulkanStateKey(typedAtom.PShaderModule.Read(ctx, a, s, nil)))
+
+	case *RecreateShaderModule:
+		addWrite(&b, g, vulkanStateKey(typedAtom.PShaderModule.Read(ctx, a, s, nil)))
+
+	case *VkCmdCopyImage:
+		addRead(&b, g, vulkanStateKey(typedAtom.SrcImage))
+		addRead(&b, g, vulkanStateKey(typedAtom.DstImage))
+		srcBindings := getOverlappedBindingsForImage(typedAtom.SrcImage)
+		dstBindings := getOverlappedBindingsForImage(typedAtom.DstImage)
+		addRead(&b, g, g.getOrCreateCommandBuffer(typedAtom.CommandBuffer).handle)
+		addModify(&b, g, g.getOrCreateCommandBuffer(typedAtom.CommandBuffer).records)
+		recordCommand(typedAtom.CommandBuffer, func(b *AtomBehaviour) {
+			readMemoryBindingsData(b, srcBindings)
+			// Be conservative here. Without tracking all the memory ranges and
+			// calculating the memory according to the copy region, we cannot assume
+			// this command overwrites the data. So it is labelled as 'modify' to
+			// kept the previous writes
+			modifyMemoryBindingsData(b, dstBindings)
+		})
+
+	case *RecreateCmdCopyImage:
+		addRead(&b, g, vulkanStateKey(typedAtom.SrcImage))
```

Comment:
I think it would make sense to refactor this a bit

srcBindings := addReadAndGetOverlappedBindings(&b, g, typedAtom.srcImage)

This pattern is used a lot, and it would save some code.